Make the model predict better! You can find additional reviews to better train the model, or you can extract additional features to include.

Write the necessary code. Explain your approach.

What can we learn from this classification task?

In [1]:
import pandas as pd
import numpy as np
import statsmodels.api as sm
from matplotlib import pyplot as plt
import seaborn as sns
import json
import math
import collections as cl

!pip install boilerpipe3
from boilerpipe.extract import Extractor 

You should consider upgrading via the '/Users/joanner/opt/anaconda3/bin/python -m pip install --upgrade pip' command.


/Users/joanner/opt/anaconda3/lib/python3.7/site-packages/jpype/_core.py:209: UserWarning: 
-------------------------------------------------------------------------------
Deprecated: convertStrings was not specified when starting the JVM. The default
behavior in JPype will be False starting in JPype 0.8. The recommended setting
for new code is convertStrings=False.  The legacy value of True was assumed for
this session. If you are a user of an application that reported this warning,
please file a ticket with the developer.
-------------------------------------------------------------------------------

  """)


In [2]:
import nltk
nltk.download('punkt')
nltk.download('averaged_perceptron_tagger')
nltk.download('maxent_ne_chunker')
nltk.download('words')
nltk.download('wordnet')

[nltk_data] Downloading package punkt to /Users/joanner/nltk_data...
[nltk_data]   Package punkt is already up-to-date!
[nltk_data] Downloading package averaged_perceptron_tagger to
[nltk_data]     /Users/joanner/nltk_data...
[nltk_data]   Package averaged_perceptron_tagger is already up-to-
[nltk_data]       date!
[nltk_data] Downloading package maxent_ne_chunker to
[nltk_data]     /Users/joanner/nltk_data...
[nltk_data]   Package maxent_ne_chunker is already up-to-date!
[nltk_data] Downloading package words to /Users/joanner/nltk_data...
[nltk_data]   Package words is already up-to-date!
[nltk_data] Downloading package wordnet to /Users/joanner/nltk_data...
[nltk_data]   Package wordnet is already up-to-date!


True

In [3]:
wnl = nltk.WordNetLemmatizer()

In [4]:
URLs = [
       'https://pitchfork.com/reviews/albums/david-bowie-changesnowbowie/',
       'https://pitchfork.com/reviews/albums/flat-worms-antarctica/',
       'https://pitchfork.com/reviews/albums/skylar-gudasz-cinema/',
       'https://pitchfork.com/reviews/albums/sir-richard-bishop-oneiric-formulary/',
       'https://pitchfork.com/reviews/albums/talking-heads-fear-of-music/',
       'https://pitchfork.com/reviews/albums/talking-heads-talking-heads-77/',
       'https://pitchfork.com/reviews/albums/talking-heads-more-songs-about-buildings-and-food/',
       'https://pitchfork.com/reviews/albums/talking-heads-speaking-in-tongues/',
       'https://pitchfork.com/reviews/albums/talking-heads-little-creatures/',
       'https://pitchfork.com/reviews/albums/the-howling-hex-knuckleball-express/',
]

rock_reviews = []
for URL in URLs:
    print(URL)
    extractor = Extractor(extractor = 'ArticleSentencesExtractor', url = URL) 
    rock_reviews.append(extractor.getText())

rock_reviews = str(rock_reviews)
rock_reviews = rock_reviews.replace('\\n', ' ')

rock_tokens = nltk.word_tokenize(rock_reviews)
rock_tokens2 = [t.lower() for t in rock_tokens if len(t) > 3] 
rock_lemma = [wnl.lemmatize(t) for t in rock_tokens2]

print(len(rock_lemma))

https://pitchfork.com/reviews/albums/david-bowie-changesnowbowie/
https://pitchfork.com/reviews/albums/flat-worms-antarctica/
https://pitchfork.com/reviews/albums/skylar-gudasz-cinema/
https://pitchfork.com/reviews/albums/sir-richard-bishop-oneiric-formulary/
https://pitchfork.com/reviews/albums/talking-heads-fear-of-music/
https://pitchfork.com/reviews/albums/talking-heads-talking-heads-77/
https://pitchfork.com/reviews/albums/talking-heads-more-songs-about-buildings-and-food/
https://pitchfork.com/reviews/albums/talking-heads-speaking-in-tongues/
https://pitchfork.com/reviews/albums/talking-heads-little-creatures/
https://pitchfork.com/reviews/albums/the-howling-hex-knuckleball-express/
5873


In [6]:
# reassigning the variables to try a different approach in addition to this one

rreviews = rock_reviews
lemrock = rock_lemma

In [7]:
rock_terms = set(['acid', 'arena', 'beat', 'garage', 'glam', 'metal', 'indie', 'prog', 'psychedelic', 'punk', 'surf', 'melody', 'soul', 'peace','heavy','wave','bounce','bulge'])
jazz_terms = set(['avant-garde', 'bebop', 'cool', 'free', 'bop', 'Latin', 'swing', 'improv', 'saxophone', 'trumpet','downtown','blue','boogie', 'woogie','ragtime','ballroom'])
rap_terms = set(['bounce', 'chopped', 'crunk', 'freestyle', 'gangsta', 'g-funk', 'trip', 'turntablism', 'mc', 'dj','hiphop', 'rap', 'battle','crossover','hit','southern','bang','south','beats','bass','snap'])
folk_terms = set(['anti-war', 'protest', 'shanty', 'bluegrass', 'cajun', 'cowboy', 'nashville', 'swamp', 'blues', 'zydeco', 'unplugged','americana','regional','ballads','country', 'appalachia','contra','square','fiddle','banjo','skiffle'])

In [8]:
URLs = [
       'https://pitchfork.com/reviews/albums/david-bowie-changesnowbowie/',
       'https://pitchfork.com/reviews/albums/flat-worms-antarctica/',
       'https://pitchfork.com/reviews/albums/skylar-gudasz-cinema/',
       'https://pitchfork.com/reviews/albums/sir-richard-bishop-oneiric-formulary/',
       'https://pitchfork.com/reviews/albums/talking-heads-fear-of-music/',
       'https://pitchfork.com/reviews/albums/talking-heads-talking-heads-77/',
       'https://pitchfork.com/reviews/albums/talking-heads-more-songs-about-buildings-and-food/',
       'https://pitchfork.com/reviews/albums/talking-heads-speaking-in-tongues/',
       'https://pitchfork.com/reviews/albums/talking-heads-little-creatures/',
       'https://pitchfork.com/reviews/albums/the-howling-hex-knuckleball-express/',
        'https://pitchfork.com/reviews/albums/the-mountain-goats-songs-for-pierre-chuvin/',
        'https://pitchfork.com/reviews/albums/nine-inch-nails-ghosts-v-together-ghosts-vi-locusts/',
        'https://pitchfork.com/reviews/albums/the-strokes-the-new-abnormal/',
        'https://pitchfork.com/reviews/albums/no-doubt-tragic-kingdom/',
        'https://pitchfork.com/reviews/albums/soccer-mommy-color-theory/',
        'https://pitchfork.com/reviews/albums/tame-impala-the-slow-rush/',
        'https://pitchfork.com/reviews/albums/green-day-father-of-all/',
        'https://pitchfork.com/reviews/albums/harry-nilsson-losst-and-found/',
       'https://pitchfork.com/reviews/albums/stay-inside-viewing/',
       'https://pitchfork.com/reviews/albums/fiona-apple-fetch-the-bolt-cutters/',
       'https://pitchfork.com/reviews/albums/ambar-lucid-garden-of-lucid/',
       'https://pitchfork.com/reviews/albums/active-child-in-another-life/',
        'https://pitchfork.com/reviews/albums/peter-gabriel-so/',
    'https://pitchfork.com/reviews/albums/american-football-year-one-demos/',
    'https://pitchfork.com/reviews/albums/st-vincent-nina-kraviz-nina-kraviz-presents-masseduction-rewired/'
]


rock_reviews = {}
for URL in URLs:
    extractor = Extractor(extractor = 'ArticleSentencesExtractor', url = URL) 
    rock_reviews[URL] = str(extractor.getText()).replace('\\n', ' ')

rock_df = pd.DataFrame()

rock_keywords = jazz_keywords = rap_keywords = folk_keywords = 0
for review in rock_reviews.values():
    rock_tokens = nltk.word_tokenize(review)
    rock_tokens2 = [t.lower() for t in rock_tokens if len(t) > 3] 
    rock_lemma = [wnl.lemmatize(t) for t in rock_tokens2]
    rock_vocab = set(rock_lemma)
    for lemma in rock_vocab:
        for term in rock_terms:
            if lemma.find(term) != -1: rock_keywords += 1
        for term in jazz_terms:
            if lemma.find(term) != -1: jazz_keywords += 1
        for term in rap_terms:
            if lemma.find(term) != -1: rap_keywords += 1
        for term in folk_terms:
            if lemma.find(term) != -1: folk_keywords += 1
    rock_freq = nltk.FreqDist(rock_lemma)
    rock_ser = pd.Series([rock_keywords, jazz_keywords, rap_keywords, folk_keywords, 
                          len(rock_vocab), len(rock_vocab)/len(rock_lemma), rock_freq.max(), 
                          len(rock_freq.hapaxes()), 'rock'], index = ['rock_keywords', 
                                                                      'jazz_keywords', 'rap_keywords', 
                                                                      'folk_keywords', 'vocab_size', 
                                                                      'lexical_diversity', 'most_common_word', 
                                                                      'number_hapaxes', 'genre'])
    rock_df = rock_df.append(rock_ser, ignore_index=True)

rock_df

folk_keywords genre  jazz_keywords  lexical_diversity most_common_word  \
0             2.0  rock            0.0           0.753695            bowie   
1             2.0  rock            0.0           0.870629       antarctica   
2             3.0  rock            1.0           0.813754           gudasz   
3             3.0  rock            3.0           0.809524           bishop   
4             4.0  rock            4.0           0.630309             that   
5             5.0  rock           10.0           0.647702             they   
6             6.0  rock           10.0           0.725694            their   
7             6.0  rock           11.0           0.743590            byrne   
8             7.0  rock           13.0           0.620112           little   
9             9.0  rock           15.0           0.777473             that   
10            9.0  rock           15.0           0.744027             with   
11            9.0  rock           16.0           0.696907             that   
12            9.0  rock           18.0           0.666667            their   
13           11.0  rock           19.0           0.632858             gwen   
14           13.0  rock           22.0           0.753125             from   
15           13.0  rock           22.0           0.701950           parker   
16           15.0  rock           24.0           0.771186            green   
17           15.0  rock           26.0           0.696226          nilsson   
18           15.0  rock           26.0           0.801829            their   
19           15.0  rock           30.0           0.644231            apple   
20           16.0  rock           31.0           0.691327            lucid   
21           16.0  rock           31.0           0.751634           grossi   
22           18.0  rock           36.0           0.613108          gabriel   
23           18.0  rock           37.0           0.718894             demo   
24           18.0  rock           38.0           0.750000     masseduction   

    number_hapaxes  rap_keywords  rock_keywords  vocab_size  
0            255.0           8.0            1.0       306.0  
1            224.0          11.0            4.0       249.0  
2            252.0          13.0            6.0       284.0  
3            296.0          14.0            7.0       340.0  
4            516.0          19.0           11.0       653.0  
5            479.0          22.0           16.0       592.0  
6            346.0          27.0           23.0       418.0  
7            388.0          33.0           28.0       464.0  
8            441.0          40.0           33.0       555.0  
9            245.0          42.0           37.0       283.0  
10           368.0          43.0           40.0       436.0  
11           255.0          44.0           43.0       338.0  
12           289.0          46.0           47.0       368.0  
13           589.0          55.0           54.0       755.0  
14           415.0          57.0           59.0       482.0  
15           406.0          58.0           70.0       504.0  
16           238.0          59.0           72.0       273.0  
17           300.0          60.0           74.0       369.0  
18           231.0          64.0           76.0       263.0  
19           533.0          69.0           79.0       670.0  
20           201.0          74.0           79.0       271.0  
21           188.0          74.0           79.0       230.0  
22           889.0          87.0           89.0      1160.0  
23           258.0          92.0           91.0       312.0  
24           251.0          93.0           94.0       291.0

In [9]:
URLs = [
       'https://pitchfork.com/reviews/albums/tom-misch-yussef-dayes-what-kinda-music/',
       'https://pitchfork.com/reviews/albums/pharoah-sanders-live-in-paris-1975/',
       'https://pitchfork.com/reviews/albums/shabaka-and-the-ancestors-we-are-sent-here-by-history/',
       'https://pitchfork.com/reviews/albums/sam-gendel-satin-doll/',
       'https://pitchfork.com/reviews/albums/talking-heads-fear-of-music/',
       'https://pitchfork.com/reviews/albums/kassa-overall-i-think-im-good/',
       'https://pitchfork.com/reviews/albums/jeremy-cunningham-the-weather-up-there/',
       'https://pitchfork.com/reviews/albums/gil-scott-heron-makaya-mccraven-were-new-again-a-reimagining-by-makaya-mccraven/',
       'https://pitchfork.com/reviews/albums/jeff-parker-suite-for-max-brown/',
       'https://pitchfork.com/reviews/albums/keith-jarrett-munich-2016/',
       'https://pitchfork.com/reviews/albums/dopolarians-garden-party/',
       'https://pitchfork.com/reviews/albums/badge-epoque-ensemble-nature-man-and-woman-ep/',
       'https://pitchfork.com/reviews/albums/junius-paul-ism/',
       'https://pitchfork.com/reviews/albums/jamael-dean-black-space-tapes/',
       'https://pitchfork.com/reviews/albums/tribe-hometown-detroit-sessions-1990-2014/',
       'https://pitchfork.com/reviews/albums/matana-roberts-coin-coin-chapter-four-memphis/',
       'https://pitchfork.com/reviews/albums/jaimie-branch-fly-or-die-ii-bird-dogs-of-paradise/',
       'https://pitchfork.com/reviews/albums/robert-glasper-fuck-yo-feelings/',
       'https://pitchfork.com/reviews/albums/kris-davis-diatom-ribbons/',
       'https://pitchfork.com/reviews/albums/the-comet-is-coming-the-afterlife/',
       'https://pitchfork.com/reviews/albums/tyshawn-sorey-marilyn-crispell-the-adornment-of-time/',
        'https://pitchfork.com/reviews/albums/john-coltrane-blue-world/',
        'https://pitchfork.com/reviews/albums/miles-davis-the-complete-birth-of-the-cool/',
        'https://pitchfork.com/reviews/albums/bennie-maupin-the-jewel-in-the-lotus/',
        'https://pitchfork.com/reviews/albums/nerija-blume/',
        'https://pitchfork.com/reviews/albums/don-cherry-brown-rice/',
        'https://pitchfork.com/reviews/albums/ray-charles-modern-sounds-in-country-and-western-music-volumes-1-and-2/'

]

jazz_reviews = {}
for URL in URLs:
    extractor = Extractor(extractor = 'ArticleSentencesExtractor', url = URL) 
    jazz_reviews[URL] = str(extractor.getText()).replace('\\n', ' ')

jazz_df = pd.DataFrame()

rock_keywords = jazz_keywords = rap_keywords = folk_keywords = 0
for review in jazz_reviews.values():
    jazz_tokens = nltk.word_tokenize(review)
    jazz_tokens2 = [t.lower() for t in jazz_tokens if len(t) > 3] 
    jazz_lemma = [wnl.lemmatize(t) for t in jazz_tokens2]
    jazz_vocab = set(jazz_lemma)
    for lemma in jazz_vocab:
        for term in rock_terms:
            if lemma.find(term) != -1: rock_keywords += 1
        for term in jazz_terms:
            if lemma.find(term) != -1: jazz_keywords += 1
        for term in rap_terms:
            if lemma.find(term) != -1: rap_keywords += 1
        for term in folk_terms:
            if lemma.find(term) != -1: folk_keywords += 1
    jazz_freq = nltk.FreqDist(jazz_lemma)
    jazz_ser = pd.Series([rock_keywords, jazz_keywords, rap_keywords, folk_keywords, 
                          len(jazz_vocab), len(jazz_vocab)/len(jazz_lemma), jazz_freq.max(), 
                          len(jazz_freq.hapaxes()), 'jazz'], index = ['rock_keywords', 'jazz_keywords', 
                                                                      'rap_keywords', 'folk_keywords', 
                                                                      'vocab_size', 'lexical_diversity', 
                                                                      'most_common_word', 'number_hapaxes', 'genre'])
    jazz_df = jazz_df.append(jazz_ser, ignore_index=True)

jazz_df

folk_keywords genre  jazz_keywords  lexical_diversity most_common_word  \
0             0.0  jazz            1.0           0.701342            dayes   
1             1.0  jazz            3.0           0.677852           sander   
2             1.0  jazz            3.0           0.711957             that   
3             2.0  jazz            6.0           0.795252           gendel   
4             3.0  jazz            7.0           0.630309             that   
5             3.0  jazz            9.0           0.750742             that   
6             4.0  jazz           12.0           0.791262       cunningham   
7             6.0  jazz           15.0           0.650909      scott-heron   
8             6.0  jazz           18.0           0.698206           parker   
9             6.0  jazz           22.0           0.753181          jarrett   
10            6.0  jazz           27.0           0.740331             with   
11           10.0  jazz           31.0           0.768535             that   
12           10.0  jazz           37.0           0.767327             paul   
13           10.0  jazz           37.0           0.787879             dean   
14           10.0  jazz           47.0           0.711354             that   
15           10.0  jazz           53.0           0.736661           robert   
16           11.0  jazz           57.0           0.777778           branch   
17           11.0  jazz           60.0           0.776382             fuck   
18           11.0  jazz           62.0           0.748837            davis   
19           11.0  jazz           62.0           0.767313             that   
20           11.0  jazz           70.0           0.775561            sorey   
21           14.0  jazz           73.0           0.650750         coltrane   
22           15.0  jazz           85.0           0.670103             cool   
23           15.0  jazz           92.0           0.616358             jazz   
24           15.0  jazz           98.0           0.743802             that   
25           16.0  jazz          105.0           0.698925           cherry   
26           21.0  jazz          111.0           0.688928          charles   

    number_hapaxes  rap_keywords  rock_keywords  vocab_size  
0            170.0           3.0            2.0       209.0  
1            306.0           5.0            6.0       404.0  
2            324.0           8.0            6.0       393.0  
3            226.0          16.0            7.0       268.0  
4            516.0          21.0           11.0       653.0  
5            425.0          32.0           15.0       506.0  
6            280.0          34.0           16.0       326.0  
7            413.0          38.0           19.0       537.0  
8            337.0          42.0           24.0       428.0  
9            244.0          46.0           24.0       296.0  
10           220.0          49.0           26.0       268.0  
11           372.0          50.0           35.0       425.0  
12           267.0          54.0           35.0       310.0  
13           244.0          56.0           38.0       286.0  
14           418.0          63.0           41.0       520.0  
15           369.0          68.0           44.0       428.0  
16           279.0          71.0           48.0       329.0  
17           268.0          74.0           51.0       309.0  
18           263.0          78.0           52.0       322.0  
19           225.0          79.0           56.0       277.0  
20           263.0          80.0           59.0       311.0  
21           386.0          83.0           61.0       477.0  
22           407.0          85.0           62.0       520.0  
23           478.0          90.0           64.0       633.0  
24           220.0          93.0           68.0       270.0  
25           315.0          95.0           73.0       390.0  
26           319.0         101.0           74.0       392.0

In [10]:
URLs = [
       'https://pitchfork.com/reviews/albums/trouble-thug-luv/',
       'https://pitchfork.com/reviews/albums/youngboy-never-broke-again-38-baby-2/',
       'https://pitchfork.com/reviews/albums/chucky73-fetti031-sie7etr3-ep/',
       'https://pitchfork.com/reviews/albums/asher-roth-flowers-on-the-weekend/',
       'https://pitchfork.com/reviews/albums/zeroh-blqlyte/',
       'https://pitchfork.com/reviews/albums/quelle-chris-innocent-country-2/',
       'https://pitchfork.com/reviews/albums/serengeti-ajai/',
       'https://pitchfork.com/reviews/albums/westside-gunn-pray-for-paris/',
       'https://pitchfork.com/reviews/albums/dababy-blame-it-on-baby/',
       'https://pitchfork.com/reviews/albums/shabazz-palaces-the-don-of-diamond-dreams/',
       'https://pitchfork.com/reviews/albums/pink-siifu-negro/',
       'https://pitchfork.com/reviews/albums/22gz-growth-and-development/',
       'https://pitchfork.com/reviews/albums/ghostie-self-hate-wraith/',
       'https://pitchfork.com/reviews/albums/rod-wave-pray-4-love/',
       'https://pitchfork.com/reviews/albums/knxwledge-1988/',
       'https://pitchfork.com/reviews/albums/bankroll-fresh-in-bank-we-trust/',
       'https://pitchfork.com/reviews/albums/ol-dirty-bastard-return-to-the-36-chambers-the-dirty-version/',
       'https://pitchfork.com/reviews/albums/rio-da-yung-og-city-on-my-back/',
       'https://pitchfork.com/reviews/albums/j-balvin-colores/',
       'https://pitchfork.com/reviews/albums/childish-gambino-31520/',
       'https://pitchfork.com/reviews/albums/zebra-katz-less-is-moor/',
    'https://pitchfork.com/reviews/albums/megan-thee-stallion-suga/',
    'https://pitchfork.com/reviews/albums/riz-ahmed-the-long-goodbye/',
    'https://pitchfork.com/reviews/albums/princess-nokia-everything-is-beautiful-everything-sucks/',
    'https://pitchfork.com/reviews/albums/ovrkast-try-again/',
    'https://pitchfork.com/reviews/albums/obongjayar-which-way-is-forward-ep/',
    'https://pitchfork.com/reviews/albums/key-glock-yellow-tape/'

]

rap_reviews = {}
for URL in URLs:
    extractor = Extractor(extractor = 'ArticleSentencesExtractor', url = URL) 
    rap_reviews[URL] = str(extractor.getText()).replace('\\n', ' ')

rap_df = pd.DataFrame()

rock_keywords = jazz_keywords = rap_keywords = folk_keywords = 0
for review in rap_reviews.values():
    rap_tokens = nltk.word_tokenize(review)
    rap_tokens2 = [t.lower() for t in rap_tokens if len(t) > 3] 
    rap_lemma = [wnl.lemmatize(t) for t in rap_tokens2]
    rap_vocab = set(rap_lemma)
    for lemma in rap_vocab:
        for term in rock_terms:
            if lemma.find(term) != -1: rock_keywords += 1
        for term in jazz_terms:
            if lemma.find(term) != -1: jazz_keywords += 1
        for term in rap_terms:
            if lemma.find(term) != -1: rap_keywords += 1
        for term in folk_terms:
            if lemma.find(term) != -1: folk_keywords += 1
    rap_freq = nltk.FreqDist(rap_lemma)
    rap_ser = pd.Series([rock_keywords, jazz_keywords, rap_keywords, folk_keywords, 
                         len(rap_vocab), len(rap_vocab)/len(rap_lemma), rap_freq.max(), 
                         len(rap_freq.hapaxes()), 'rap'], index = ['rock_keywords', 'jazz_keywords', 
                                                                   'rap_keywords', 'folk_keywords', 
                                                                   'vocab_size', 'lexical_diversity', 
                                                                   'most_common_word', 'number_hapaxes', 'genre'])
    rap_df = rap_df.append(rap_ser, ignore_index=True)

rap_df

folk_keywords genre  jazz_keywords  lexical_diversity most_common_word  \
0             0.0   rap            1.0           0.748072             that   
1             0.0   rap            2.0           0.741100         youngboy   
2             0.0   rap            2.0           0.733542             crew   
3             1.0   rap            4.0           0.806763             roth   
4             1.0   rap            7.0           0.826220            zeroh   
5             2.0   rap            8.0           0.787115            chris   
6             2.0   rap            9.0           0.749169             ajai   
7             2.0   rap            9.0           0.744966             gunn   
8             2.0   rap           10.0           0.740230           dababy   
9             2.0   rap           12.0           0.765182             that   
10            2.0   rap           15.0           0.719269             that   
11            2.0   rap           15.0           0.709239             22gz   
12            3.0   rap           19.0           0.767033          ghostie   
13            5.0   rap           20.0           0.809677             with   
14            5.0   rap           23.0           0.794788        knxwledge   
15            5.0   rap           23.0           0.748315         bankroll   
16            6.0   rap           27.0           0.664869            dirty   
17            6.0   rap           28.0           0.798762             that   
18            6.0   rap           28.0           0.819753           balvin   
19            8.0   rap           29.0           0.681458             that   
20            9.0   rap           32.0           0.764967             katz   
21           10.0   rap           34.0           0.769565            megan   
22           11.0   rap           35.0           0.743155             with   
23           12.0   rap           36.0           0.797701             that   
24           12.0   rap           38.0           0.792000          ovrkast   
25           12.0   rap           38.0           0.813953             with   
26           14.0   rap           39.0           0.810811            glock   

    number_hapaxes  rap_keywords  rock_keywords  vocab_size  
0            240.0           8.0            3.0       291.0  
1            184.0          12.0            3.0       229.0  
2            189.0          14.0            6.0       234.0  
3            288.0          18.0            6.0       334.0  
4            238.0          22.0            9.0       271.0  
5            236.0          28.0           11.0       281.0  
6            378.0          36.0           14.0       451.0  
7            286.0          44.0           16.0       333.0  
8            272.0          49.0           17.0       322.0  
9            159.0          52.0           17.0       189.0  
10           366.0          58.0           23.0       433.0  
11           220.0          64.0           27.0       261.0  
12           297.0          68.0           29.0       349.0  
13           218.0          72.0           33.0       251.0  
14           212.0          75.0           34.0       244.0  
15           279.0          84.0           38.0       333.0  
16           779.0          97.0           40.0       986.0  
17           217.0         102.0           41.0       258.0  
18           296.0         108.0           45.0       332.0  
19           335.0         112.0           46.0       430.0  
20           300.0         120.0           48.0       345.0  
21           302.0         129.0           49.0       354.0  
22           465.0         137.0           52.0       570.0  
23           300.0         139.0           53.0       347.0  
24           259.0         142.0           57.0       297.0  
25           241.0         143.0           61.0       280.0  
26           209.0         151.0           62.0       240.0

In [11]:
URLs = [
       'https://pitchfork.com/reviews/albums/lucinda-williams-good-souls-better-angels/',
       'https://pitchfork.com/reviews/albums/james-elkington-ever-roving-eye/',
       'https://pitchfork.com/reviews/albums/laura-marling-song-for-our-daughter/',
       'https://pitchfork.com/reviews/albums/mosses-tv-sun/',
       'https://pitchfork.com/reviews/albums/sam-hunt-southside/',
       'https://pitchfork.com/reviews/albums/william-tyler-music-from-first-cow/',
       'https://pitchfork.com/reviews/albums/brandy-clark-your-life-is-a-record/',
       'https://pitchfork.com/reviews/albums/katie-pruitt-expectations/',
       'https://pitchfork.com/reviews/albums/john-moreland-lp5/',
       'https://pitchfork.com/reviews/albums/bill-fay-countless-branches/',
       'https://pitchfork.com/reviews/albums/aoife-nessa-frances-land-of-no-junction/',
       'https://pitchfork.com/reviews/albums/loving-if-i-am-only-my-thoughts/',
       'https://pitchfork.com/reviews/albums/terry-allen-and-the-panhandle-mystery-ba-just-like-moby-dick/',
       'https://pitchfork.com/reviews/albums/dixie-chicks-home/',
       'https://pitchfork.com/reviews/albums/string-machine-death-of-the-neon/',
       'https://pitchfork.com/reviews/albums/sarah-mary-chadwick-please-daddy/',
       'https://pitchfork.com/reviews/albums/lina-raul-refree-lina-raul-refree/',
       'https://pitchfork.com/reviews/albums/friendship-dreamin/',
       'https://pitchfork.com/reviews/albums/san-cha-la-luz-de-la-esperanza/',
       'https://pitchfork.com/reviews/albums/grace-cummings-refuge-cove/',
       'https://pitchfork.com/reviews/albums/hurt-valley-glacial-pace/',
    'https://pitchfork.com/reviews/albums/konradsen-saints-and-sebastian-stories/',
    'https://pitchfork.com/reviews/albums/sufjan-stevens-timo-andres-the-decalogue/',
    'https://pitchfork.com/reviews/albums/kd-lang-ingenue/',
    'https://pitchfork.com/reviews/albums/patsy-cline-sentimentally-yours/',
    'https://pitchfork.com/reviews/albums/lil-nas-x-7-ep/',
    'https://pitchfork.com/reviews/albums/calexico-iron-and-wine-years-to-burn/'

]

folk_reviews = {}
for URL in URLs:
    extractor = Extractor(extractor = 'ArticleSentencesExtractor', url = URL) 
    folk_reviews[URL] = str(extractor.getText()).replace('\\n', ' ')

folk_df = pd.DataFrame()

rock_keywords = jazz_keywords = rap_keywords = folk_keywords = 0
for review in folk_reviews.values():
    folk_tokens = nltk.word_tokenize(review)
    folk_tokens2 = [t.lower() for t in folk_tokens if len(t) > 3] 
    folk_lemma = [wnl.lemmatize(t) for t in folk_tokens2]
    folk_vocab = set(folk_lemma)
    for lemma in folk_vocab:
        for term in rock_terms:
            if lemma.find(term) != -1: rock_keywords += 1
        for term in jazz_terms:
            if lemma.find(term) != -1: jazz_keywords += 1
        for term in rap_terms:
            if lemma.find(term) != -1: rap_keywords += 1
        for term in folk_terms:
            if lemma.find(term) != -1: folk_keywords += 1
    folk_freq = nltk.FreqDist(folk_lemma)
    folk_ser = pd.Series([rock_keywords, jazz_keywords, rap_keywords, folk_keywords, 
                          len(folk_vocab), len(folk_vocab)/len(folk_lemma), folk_freq.max(), 
                          len(folk_freq.hapaxes()), 'folk'], index = ['rock_keywords', 'jazz_keywords', 
                                                                      'rap_keywords', 'folk_keywords', 
                                                                      'vocab_size', 'lexical_diversity', 
                                                                      'most_common_word', 'number_hapaxes', 'genre'])
    folk_df = folk_df.append(folk_ser, ignore_index=True)

folk_df

folk_keywords genre  jazz_keywords  lexical_diversity most_common_word  \
0             5.0  folk            2.0           0.724138         williams   
1             5.0  folk            3.0           0.748705        elkington   
2             6.0  folk            6.0           0.738255          marling   
3             6.0  folk            7.0           0.803922             that   
4            13.0  folk           10.0           0.733333          country   
5            18.0  folk           11.0           0.774194             with   
6            20.0  folk           11.0           0.738866             with   
7            22.0  folk           12.0           0.751412             with   
8            24.0  folk           12.0           0.808219         moreland   
9            25.0  folk           12.0           0.756494            album   
10           25.0  folk           13.0           0.805158           france   
11           25.0  folk           13.0           0.763819             that   
12           27.0  folk           14.0           0.789130            allen   
13           33.0  folk           19.0           0.560129            their   
14           35.0  folk           21.0           0.787013           string   
15           35.0  folk           22.0           0.757143             that   
16           35.0  folk           23.0           0.745636           refree   
17           36.0  folk           23.0           0.775610             that   
18           37.0  folk           23.0           0.746803             with   
19           37.0  folk           23.0           0.766447             that   
20           37.0  folk           23.0           0.774194          collins   
21           38.0  folk           24.0           0.794953            music   
22           39.0  folk           25.0           0.770308          stevens   
23           41.0  folk           27.0           0.718588             lang   
24           50.0  folk           29.0           0.636196             that   
25           52.0  folk           29.0           0.705495             that   
26           53.0  folk           31.0           0.707379             that   

    number_hapaxes  rap_keywords  rock_keywords  vocab_size  
0            350.0           5.0            5.0       420.0  
1            250.0           5.0            8.0       289.0  
2            277.0           7.0           12.0       330.0  
3            292.0           9.0           17.0       328.0  
4            308.0          13.0           20.0       374.0  
5            259.0          14.0           20.0       312.0  
6            304.0          15.0           20.0       365.0  
7            222.0          16.0           21.0       266.0  
8            306.0          17.0           25.0       354.0  
9            193.0          17.0           28.0       233.0  
10           239.0          17.0           32.0       281.0  
11           254.0          20.0           36.0       304.0  
12           314.0          26.0           37.0       363.0  
13           654.0          32.0           44.0       871.0  
14           262.0          35.0           47.0       303.0  
15           209.0          36.0           49.0       265.0  
16           252.0          36.0           51.0       299.0  
17           268.0          36.0           53.0       318.0  
18           234.0          37.0           55.0       292.0  
19           185.0          38.0           57.0       233.0  
20           175.0          40.0           60.0       216.0  
21           209.0          40.0           60.0       252.0  
22           235.0          42.0           61.0       275.0  
23           578.0          44.0           64.0       692.0  
24           540.0          54.0           69.0       689.0  
25           265.0          60.0           74.0       321.0  
26           217.0          61.0           76.0       278.0

In [12]:
reviews_df = pd.concat([rock_df, jazz_df, rap_df, folk_df])
reviews_df

folk_keywords genre  jazz_keywords  lexical_diversity most_common_word  \
0             2.0  rock            0.0           0.753695            bowie   
1             2.0  rock            0.0           0.870629       antarctica   
2             3.0  rock            1.0           0.813754           gudasz   
3             3.0  rock            3.0           0.809524           bishop   
4             4.0  rock            4.0           0.630309             that   
..            ...   ...            ...                ...              ...   
22           39.0  folk           25.0           0.770308          stevens   
23           41.0  folk           27.0           0.718588             lang   
24           50.0  folk           29.0           0.636196             that   
25           52.0  folk           29.0           0.705495             that   
26           53.0  folk           31.0           0.707379             that   

    number_hapaxes  rap_keywords  rock_keywords  vocab_size  
0            255.0           8.0            1.0       306.0  
1            224.0          11.0            4.0       249.0  
2            252.0          13.0            6.0       284.0  
3            296.0          14.0            7.0       340.0  
4            516.0          19.0           11.0       653.0  
..             ...           ...            ...         ...  
22           235.0          42.0           61.0       275.0  
23           578.0          44.0           64.0       692.0  
24           540.0          54.0           69.0       689.0  
25           265.0          60.0           74.0       321.0  
26           217.0          61.0           76.0       278.0  

[106 rows x 9 columns]

In [14]:
model_data_Y = reviews_df['genre']
model_data_X = reviews_df[['lexical_diversity', 'rock_keywords', 'jazz_keywords', 'rap_keywords', 'folk_keywords']]

from sklearn.model_selection import train_test_split
X_train, X_test, y_train, y_test = train_test_split(
    model_data_X, model_data_Y, test_size=0.3, random_state=0
)

from sklearn.svm import LinearSVC
model0 = LinearSVC(C = 0.01, max_iter = 1000000).fit(X_train, y_train)
model0_pred = model0.predict(X_test)

In [16]:
from sklearn.metrics import classification_report, confusion_matrix

In [19]:
print(classification_report(y_test, model0_pred))

              precision    recall  f1-score   support

        folk       1.00      1.00      1.00         8
        jazz       1.00      0.60      0.75         5
         rap       0.90      1.00      0.95         9
        rock       0.91      1.00      0.95        10

    accuracy                           0.94        32
   macro avg       0.95      0.90      0.91        32
weighted avg       0.94      0.94      0.93        32



In [20]:
model_data_Y = reviews_df['genre']
model_data_X = reviews_df[['vocab_size', 'rock_keywords', 'jazz_keywords', 'rap_keywords', 'folk_keywords']]

from sklearn.model_selection import train_test_split
X_train, X_test, y_train, y_test = train_test_split(
    model_data_X, model_data_Y, test_size=0.3, random_state=0
)

from sklearn.svm import LinearSVC
model0 = LinearSVC(C = 0.01, max_iter = 1000000).fit(X_train, y_train)
model0_pred = model0.predict(X_test)

print(classification_report(y_test, model0_pred))

              precision    recall  f1-score   support

        folk       1.00      1.00      1.00         8
        jazz       1.00      0.80      0.89         5
         rap       0.89      0.89      0.89         9
        rock       0.91      1.00      0.95        10

    accuracy                           0.94        32
   macro avg       0.95      0.92      0.93        32
weighted avg       0.94      0.94      0.94        32



In [21]:
model_data_Y = reviews_df['genre']
model_data_X = reviews_df[['vocab_size', 'rock_keywords', 'jazz_keywords', 'rap_keywords', 'folk_keywords']]

from sklearn.model_selection import train_test_split
X_train, X_test, y_train, y_test = train_test_split(
    model_data_X, model_data_Y, test_size=0.5, random_state=0
)

from sklearn.svm import LinearSVC
model0 = LinearSVC(C = 0.01, max_iter = 1000000).fit(X_train, y_train)
model0_pred = model0.predict(X_test)

print(classification_report(y_test, model0_pred))

              precision    recall  f1-score   support

        folk       1.00      1.00      1.00        12
        jazz       0.79      0.92      0.85        12
         rap       0.81      1.00      0.90        13
        rock       1.00      0.69      0.81        16

    accuracy                           0.89        53
   macro avg       0.90      0.90      0.89        53
weighted avg       0.91      0.89      0.88        53



By simply adding five more reviews to each dataset for a total of 106 records, we get a much higher accuracy rate in the classification model. Adding more records to the training set will usually give you a higher accuracy rate due to the model simply having more reference to work with--a little like being a child versus being an adult. If you don't have experience to fall back on, you won't be as confident in what you're doing. 

Regarding what we can learn from this classification task, it's clear that if the test size is increased, the support increases as well. Considering that some of these albums are tagged in multiple categories (consider Lil Nas X as being classified in both rap and folk/country), it's not surprising that the accuracy levels would be so high--if boundaries are not as strict, of course you'd be right more of the time.